drivers: clock: stm32: fix Flash latency & clock settings for MSI & HSE

giansta opened the pull request:
According to RM, when increasing the CPU frequency, the new number of
wait states to the Flash latency bits must be written and verified before
modifying the CPU clock source and/or the CPU clock prescaler; when
decreasing the CPU frequency, after.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -381,6 +387,22 @@ static int stm32_clock_control_init(struct device *dev)
+
+	/* If freq increases, set flash latency before any clock setting */
+	if (new_hclk_freq > old_hclk_freq) {
+		LL_SetFlashLatency(new_hclk_freq);
```
Unfortunately, today this function is only available for a small set of series (L4, L5, G4).
For other series it is available under its private version: UTILS_SetFlashLatency. They will be progressively converted to the public version, but this is not done yet. 
What you can do is to update hal stm32 module to get all the series converted to the new public API.

giansta replied:
Thanks, @erwango 
Indeed I realized that the SetFlashLatency function was available only for L4/L5/G4, and I also had a look to the RM of the WB55 series, for which I got the fails, and the flash latencies management looked so close to the L4 series we're using... Even setting up a project for the WB55 using Stm32CubeIde confirmed that the interfaces were very close.
But your hint is now welcome, I hadn't realized yet that the other series had the same function as private. Tomorrow I'll work at this :-)

erwango reviewed (changes requested):
Thanks for this change but as of now it's not compatible with all stm32 series.
You can either limit to the series that enjoy the SetFlashLatency in public API (but this may lead to hard to read code) or update hal modules waiting for the conversion to be delivered in cube packages (on going effort).

erwango reviewed (commented):
Please rebase following merge of https://github.com/zephyrproject-rtos/zephyr/pull/25679.
Also, you can move this from daft to usual PR

erwango commented:
@giansta can you rebase on latest master to see if this gets buildkite happy ?

giansta commented:
> 
> 
> @giansta can you rebase on latest master to see if this gets buildkite happy ?
HI @erwango As I expected, both shippable and buildkite are now failing on STM32WB. This PR depends on hal_stm32 PR #54, that's the reason why it was changed into draft status. Also the hal_stm32 revision was pointing to pull/54/head that lately was conflicting with 6812113.

Commit:
drivers: clock: stm32: fix Flash latency & clock settings for MSI & HSE

According to RM, when increasing the CPU frequency, the new number of
wait states to the Flash latency bits must be written and verified
before modifying the CPU clock source and/or the CPU clock prescaler,
to prevent NMI to occur; when decreasing the CPU frequency, after.
Tested with STM32L462 SOC and MSI with several frequencies, both
increasing and decreasing. HSE built, not tested.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi@nexxiot.com>

erwango commented:
> HI @erwango As I expected, both shippable and buildkite are now failing on STM32WB

This is not what I'm seeing. At least on shippable side, error is not related to this PR.
Anyway, they should be passing even if there is a dependency with hal_stm32 (both Shippable and buildkite are able to handle this).

erwango commented:
Closing/reopenning to get new CI status.

giansta commented:
> 
> 
> Closing/reopenning to get new CI status.

Thanks.
In second-last rebase I was seeing fails with STM32W, because I had removed the revision pull/54/head, as there was a conflict with another change with hal_stm32 revision.
Then I tried the last rebase, re-adding revision pull/54/head, and I got build errors in other parts.  With Shippable I saw error (e.g. mem_pool_threadsafe) that don't seem to be related with this PR (previously all tests were passing).
With buildkite I don't understand where the error is, that is again the same now.

erwango commented:
> With buildkite I don't understand where the errors are, the output seems to be again the same now.

I don't get them neither.
Can you have a try and rebase on latest master?

giansta commented:
> 
> 
> > With buildkite I don't understand where the errors are, the output seems to be again the same now.
> 
> I don't get them neither.
> Can you have a try and rebase on latest master?

Now Shippable tests were all passed. Anyway, rebased another time ;-)

giansta commented:
@erwango Now it's much better, builkite has only one fail, yet it looks like a kernel unit test fail on qemu shouldn't be caused by a fix in STM32 driver.

ABOSTM reviewed (approved):
LGTM

tagunil commented:
Why was this PR merged before corresponding hal_stm32 PR?

pabigot commented:
Agreed.  #26424 to track this.
